src/io/network/addrinfo.hpp

```diff
@@ -11,21 +11,39 @@
 
 #pragma once
 
+#include <netdb.h>
+#include <string>
+
 namespace io::network {
 
+struct Endpoint;
```

Comment:
That means we start to use  forward declarations! FYI: it is not common in the codebase, but I am totally up for it.

Reply:
PLz don't unless required :D

Reply:
Forward declarations reduce compile time, but if it's not common or you guys don't like it I can use `#include` instead :smile:

Reply:
Hm, interesting... clang-format reorders the include files in the cpp file, which in turn can cause problems with the forward declared identifiers, i.e. if the include file that contains the forward declaration is included before the actual declaration.

Reply:
I would need a benchmark for the argument  "Forward decls reduce compile time". There are so much more to optimize out at compile time that I hardly believe this would make a difference :D

Reply:
In 99% percent of the time (assuming forward declarations are used consistently), forward declarations are faster. Check [this SO answer](https://stackoverflow.com/a/61901220).

Reply:
> It generates 200 headers. Each header has a struct with 100 fields and a comment 5000 bytes long.
I am not convinced.

How does this represent a codebase? What are the dependencies between the headers? Which file includes what? How similar is our codebase to this example (basically how's our dependency DAG vs the one in the example)? This is an overly generalized benchmark/example that does not have any meaningful information or relation to an actual codebase. 

Moreover, there are two nice answers above, and one hint about why forward decls decrease compilation time (the source is John Lakos book). However, that book is old and compilers have changed significantly since then which makes me question if that's still the case. 

All in all, unless forward decls provably solve your builds compile time for a given codebase (that is you benchmarked and you actually proven that the drag on the builds is not from SFINAE or other compile intensive computations), I would stick away from that as I don't think they make your code more readable (quite the opposite). And even then. there are other solutions, precompiled headers, and C++20 modules :P